File: lang/go/cxlib/atom.go
// atom.go atom scalar kind for the Go binding.
//
// Atom is a name-equality, type-strict scalar kind ("symbol" in other
// ecosystems). Surface form `:NAME`. Type-strict equality, Layer-1
// surface, round-trip + hash domain, and a closed reserved-name list.
//
// The Go binding's Layer-1 surface matches V's atom(name) / is_atom(v) /
// atom_name(v) triple, adapted to Go's PascalCase convention per
// `spec/bindings.md`:
//
//   cx.Atom(name)    — constructor returning an AtomValue
//   cx.IsAtom(v)     — predicate; type-strict (strings return false)
//   cx.AtomName(v)   — accessor; non-atom returns "" + false
//   v.AtomName()     — method accessor on AtomValue
//
// Atom values flow through Attr.Value / ScalarNode.Value as the typed
// AtomValue wrapper (rather than a bare string) so equality is naturally
// type-strict: an AtomValue compares unequal to a same-named string by
// Go's standard interface-equality rules.

package cxlib

import (
	"fmt"
	"regexp"
)

// ── AtomValue ────────────────────────────────────────────────────────────────

// AtomValue is the typed wrapper carrying an atom's name through the
// AST. Equality via Go's `==` and `reflect.DeepEqual` is by Name only,
// and an AtomValue compares unequal to any plain string with the same
// content — that's the type-strict equality rule.
//
// The struct is value-typed (not a pointer) so it behaves as a scalar
// at every callsite. It's also comparable, so it can serve as a map key
// inside Go (note: forbids atoms as keys in CX Map nodes;
// that rule applies to the on-the-wire shape, not Go internals).
type AtomValue struct {
	Name string
}

// String implements fmt.Stringer with the canonical render `:name`
func (a AtomValue) String() string { return ":" + a.Name }

// AtomName returns the atom's name (without the leading colon). Method
// accessor per spec/bindings.md §Layer 1 — Go uses receiver methods
// where idiomatic.
func (a AtomValue) AtomName() string { return a.Name }

// ── Layer-1 surface ───────────────────────────────────────────

// atomReservedNames is the closed-list reservation. Three
// names that would shadow scalar literals (`:true` / `:false` / `:null`)
// are rejected at lex time in V (vcx/code/parser.v parse_atom_literal).
// The Go binding mirrors the rule at construction time.
var atomReservedNames = map[string]struct{}{
	"true":  {},
	"false": {},
	"null":  {},
}

// atomNameRe matches the identifier production from spec/grammar.ebnf §3.4
// `[A-Za-z_][A-Za-z0-9_-]*`.
var atomNameRe = regexp.MustCompile(`^[A-Za-z_][A-Za-z0-9_-]*$`)

// Atom constructs a typed atom value, validating against the
// §D1 identifier production and §D8 reserved-name list. Returns an
// error rather than panicking so callers can decide between fail-fast
// (`MustAtom`) and graceful degradation.
func Atom(name string) (AtomValue, error) {
	if _, reserved := atomReservedNames[name]; reserved {
		return AtomValue{}, fmt.Errorf(
			"atom literal ':%s' is reserved; "+
				"use bare '%s' for the bool/null scalar", name, name)
	}
	if !atomNameRe.MatchString(name) {
		return AtomValue{}, fmt.Errorf(
			"atom name %q does not match identifier production "+
				"[A-Za-z_][A-Za-z0-9_-]*", name)
	}
	return AtomValue{Name: name}, nil
}

// MustAtom is the panic-on-invalid variant of Atom, suitable for
// compile-time-known constants (`var OK = cx.MustAtom("ok")`).
func MustAtom(name string) AtomValue {
	a, err := Atom(name)
	if err != nil {
		panic(err)
	}
	return a
}

// IsAtom returns true iff v is an AtomValue. Per §D2
// equality is type-strict — strings, ints, and other scalar kinds all
// return false even if their content matches an atom's name.
func IsAtom(v any) bool {
	_, ok := v.(AtomValue)
	return ok
}

// AtomName returns the name of an AtomValue and true; for non-atom
// values it returns "" and false. The dual-return shape matches Go's
// type-assertion idiom (cf. `v, ok := m[k]`) and avoids the panic
// surface of a single-return accessor.
//
// To get the atom name as a method instead, use `v.AtomName()` on a
// typed AtomValue receiver.
func AtomName(v any) (string, bool) {
	if a, ok := v.(AtomValue); ok {
		return a.Name, true
	}
	return "", false
}

File: lang/rust/cxlib/src/atom.rs
//! atom scalar kind for the Rust binding.
//!
//! Atom is a name-equality, type-strict scalar kind ("symbol" in other
//! ecosystems). Surface form `:NAME`. See
//! §D2 (type-strict
//! equality), §D5 (Layer-1 surface), §D7 (round-trip + hash domain),
//! §D8 (reserved-name closed list), §D10 (gate matrix).
//!
//! The Rust binding's atom layer maps the on-the-wire CX representation
//! (`Node::Scalar { data_type: "atom", value: Value::String(name) }`) to
//! a typed `Atom` newtype wrapper at the Layer-1 surface. The AST node
//! itself keeps `serde_json::Value` so existing accessors and the
//! ast_bin codec don't need a sum-type bump; the `is_atom` /
//! `atom_name` helpers project the `(data_type, value)` pair into a
//! type-strict view.
//!
//! Layer-1 surface (matches V's `atom(name)` / `is_atom(v)` /
//! `atom_name(v)` triple, adapted to Rust's snake_case convention per
//! `spec/bindings.md` and Rust idioms):
//!
//! ```ignore
//! use cxlib::Atom;
//! let a = Atom::new("ok")?;
//! assert_eq!(a.name(), "ok");
//! assert_eq!(a.to_string(), ":ok");      // canonical render (§D7)
//! assert!(a.as_node_pair().0 == "atom");
//! ```

use std::fmt;

use serde_json::Value;

// ── reserved-name list ────────────────────────────────────────

/// closed-list reservation. The three names that would
/// shadow scalar literals (`:true` / `:false` / `:null`) are rejected
/// at lex time in V (`vcx/code/parser.v` `parse_atom_literal`); the
/// Rust binding mirrors the rule at construction time.
const RESERVED_ATOM_NAMES: &[&str] = &["true", "false", "null"];

/// Identifier production from `spec/grammar.ebnf §3.4` — atoms accept
/// `[A-Za-z_][A-Za-z0-9_-]*`. Hand-rolled rather than depending on the
/// regex crate; the production is simple enough.
fn is_valid_atom_name(name: &str) -> bool {
    let mut chars = name.chars();
    let first = match chars.next() {
        Some(c) => c,
        None => return false,
    };
    if !(first.is_ascii_alphabetic() || first == '_') {
        return false;
    }
    for c in chars {
        if !(c.is_ascii_alphanumeric() || c == '_' || c == '-') {
            return false;
        }
    }
    true
}

// ── Atom newtype ─────────────────────────────────────────────────────────────

/// Type-strict atom scalar value.
///
/// Equality is by `name` only, and an `Atom` compares unequal to any
/// plain `&str` / `String` with the same content — that's the
/// type-strict equality rule of §D2. The type is `Clone + PartialEq +
/// Eq + Hash` so it slots cleanly into Rust collections.
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct Atom {
    name: String,
}

impl Atom {
    /// Construct a typed atom, validating against 
    /// (identifier production) and §D8 (reserved-name list).
    ///
    /// Returns `Err` rather than panicking so callers can decide
    /// between fail-fast (`Atom::new(...).unwrap()`) and graceful
    /// degradation.
    pub fn new(name: impl Into<String>) -> Result<Self, String> {
        let name = name.into();
        if RESERVED_ATOM_NAMES.contains(&name.as_str()) {
            return Err(format!(
                "atom literal ':{}' is reserved; \
                 use bare '{}' for the bool/null scalar",
                name, name
            ));
        }
        if !is_valid_atom_name(&name) {
            return Err(format!(
                "atom name {:?} does not match identifier production \
                 [A-Za-z_][A-Za-z0-9_-]*",
                name
            ));
        }
        Ok(Atom { name })
    }

    /// Construct without validation. Use only when the name is already
    /// known to be well-formed (e.g., reconstructing from a verified
    /// `ast_bin` decode). The public-API `new` is the safe path.
    pub(crate) fn from_decoded(name: String) -> Self {
        Atom { name }
    }

    /// The atom's name (without the leading colon).
    pub fn name(&self) -> &str {
        &self.name
    }

    /// Project to the `(data_type, value)` pair the AST uses on the wire.
    /// The wire form is `Scalar(data_type="atom",
    /// value=String(name))`. Useful when building `Node::Scalar`
    /// fragments by hand.
    pub fn as_node_pair(&self) -> (&'static str, Value) {
        ("atom", Value::String(self.name.clone()))
    }
}

impl fmt::Display for Atom {
    /// Canonical render — `:NAME`. No quoting, no
    /// bracketing, byte-identical to source text.
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, ":{}", self.name)
    }
}

// ── Layer-1 free functions ───────────────────────────────────────────────────

/// Predicate over a `Node::Scalar`'s `(data_type, value)` pair —
/// returns `true` iff the pair represents an atom on the wire. Per
/// §D2 the predicate is type-strict; a `Value::String` with
/// `data_type=None` (i.e., a regular string) returns false, and a
/// well-tagged `data_type=Some("atom")` with a non-string `value`
/// (malformed payload) returns false too — both halves of the wire
/// shape must agree.
pub fn is_atom(data_type: Option<&str>, value: &Value) -> bool {
    data_type == Some("atom") && matches!(value, Value::String(_))
}

/// Accessor returning the atom name as a borrowed `&str`, or `None` if
/// the `(data_type, value)` pair is not an atom. Type-strict per §D2 —
/// no coercion from a same-named string.
pub fn atom_name<'a>(data_type: Option<&str>, value: &'a Value) -> Option<&'a str> {
    if data_type == Some("atom") {
        match value {
            Value::String(s) => Some(s.as_str()),
            _ => None,
        }
    } else {
        None
    }
}
